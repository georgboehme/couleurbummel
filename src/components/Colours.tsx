import {makeStyles} from '@rneui/themed';
import {memo} from 'react';
import {ColorValue, View, ViewStyle} from 'react-native';

interface ColoursProps {
  colours?: ColorValue[];
  percussion?: ColorValue;
  baseColour?: ColorValue;
  coloursKey: string;
}

/**
 * Represents a sash or flag-like collection of colours, such as a tricolore
 */
export const Colours = memo(
  ({colours, percussion, baseColour, coloursKey}: ColoursProps) => {
    const styles = useStyles();

    if (!colours) {
      return null;
    }

    let mainStyle = [styles.mainContainer as ViewStyle];

    if (percussion) {
      mainStyle.push(styles.percussionPadding, {
        backgroundColor: percussion,
      });
    } else if (baseColour) {
      mainStyle.push(styles.baseColourPadding, {
        backgroundColor: baseColour,
      });
    }

    return (
      <View style={styles.container}>
        <View style={mainStyle}>
          {colours.map((col, i) => (
            <View
              key={`${coloursKey}_${i}`}
              style={[styles.colour, {backgroundColor: col}]}
            />
          ))}
        </View>
      </View>
    );
  },
);

const useStyles = makeStyles(theme => ({
  container: {
    aspectRatio: 3 / 2,
  },
  mainContainer: {
    justifyContent: 'center',
    borderRadius: 5,
    overflow: 'hidden',
    borderColor: theme.colors.greyOutline,
    borderWidth: 1,
    width: '100%',
    flexGrow: 1,
    flexShrink: 1,
  },
  colour: {
    flex: 1,
    flexDirection: 'column',
    width: '100%',
  },
  percussionPadding: {
    paddingTop: '5%',
    paddingBottom: '5%',
  },
  baseColourPadding: {
    paddingTop: '12%',
    paddingBottom: '12%',
  },
}));

export default Colours;
