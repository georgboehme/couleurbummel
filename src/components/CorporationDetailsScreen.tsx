import {makeStyles} from '@rneui/themed';
import React, {memo, useCallback, useEffect} from 'react';

import {CorporationDetailsScreenProps} from '../types/screens';
import {CorporationDetails} from './details/Corporation';
import {RequestChangeButton} from './details/RequestChangeButton';

/**
 * A screen displaying the details of a corporation
 */
export const CorporationDetailsScreen = memo(
  ({navigation, route}: CorporationDetailsScreenProps) => {
    const styles = useStyles();
    const corporation = route.params.corporation;

    const RequestChangeForCorporationButton = useCallback(
      () => <RequestChangeButton name={corporation.shortName} />,
      [corporation],
    );

    useEffect(() => {
      navigation.setOptions({
        title: corporation.shortName,
        headerRight: RequestChangeForCorporationButton,
      });
    });

    return (
      <CorporationDetails
        corporations={[corporation]}
        containerStyle={styles.container}
      />
    );
  },
);

const useStyles = makeStyles(() => ({
  container: {
    height: '100%',
  },
}));
