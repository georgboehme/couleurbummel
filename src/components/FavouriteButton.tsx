import {Button, makeStyles, useTheme} from '@rneui/themed';
import React, {memo, useContext, useEffect, useState} from 'react';
import {useTranslation} from 'react-i18next';

import constants from '../constants';
import log from '../log';
import {ToggleFavouriteAction} from '../types/state';
import {StateContext} from './contexts/GlobalStateContext';

export interface FavouriteButtonProps {
  itemId: string;
  toggleAction: (id: string) => ToggleFavouriteAction;
  postDispatch?: () => void;
}

/**
 * A button to toggle an item as a favourite
 *
 * @param itemId the ID of the item
 * @param toggleAction the action to dispatch to toggle the item as a favourite
 * @param postDispatch function to call after dispatching
 */
export const FavouriteButton = memo(
  ({itemId, toggleAction, postDispatch}: FavouriteButtonProps) => {
    const {globalState, dispatch} = useContext(StateContext);
    const styles = useStyles();
    const {theme} = useTheme();
    const {t} = useTranslation();

    const [isFavourite, setIsFavourite] = useState(
      globalState.favourites.corporation.has(itemId),
    );

    useEffect(() => {
      log.debug('State changed: FavouriteButton');
      setIsFavourite(globalState.favourites.corporation.has(itemId));
    }, [globalState, itemId]);

    return (
      <Button
        buttonStyle={styles.favouriteButton}
        title={t('BUTTON_FAVOURITE') || ''}
        icon={{
          name: isFavourite
            ? constants.icons.explore.favourite
            : constants.icons.explore.nonFavourite,
          color: theme.colors.iconOnPrimaryColourBackgroundColour,
        }}
        onPress={() => {
          dispatch(toggleAction(itemId));
          if (postDispatch) {
            postDispatch();
          }
        }}
      />
    );
  },
);

const useStyles = makeStyles(() => ({
  favouriteButton: {
    height: '100%',
  },
}));
