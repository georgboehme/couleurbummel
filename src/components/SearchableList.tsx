import {makeStyles, SearchBar, useTheme} from '@rneui/themed';
import React, {useMemo, useState} from 'react';
import {useTranslation} from 'react-i18next';
import {FlatList, Platform, View} from 'react-native';

import constants from '../constants';
import log from '../log';

export interface SearchableListProps<T> {
  allItems: T[];
  filterItems: (searchTerm: string, items: T[]) => T[];
  renderItem: (item: T) => React.ReactElement;
  keyExtractor: (item: T) => string;
  showsAllResultsOnEmptySearch?: boolean;
  searchDelayMs?: number;
  headerComponent?: React.ReactElement<{}>;
  testID?: string;
}

/**
 * A flat list with a search bar
 *
 * @param filterItems function filtering the items based on the search text
 * @param allItems all possible items
 * @param renderItem function rendering the row for a given item
 * @param keyExtractor function returning the key for an item
 * @param searchDelayMs how long to wait for another key press before starting the search
 * @param testID a string to identify the list in a test setting
 * @param showsAllResultsOnEmptySearch if true and the search bar is empty, all results are shown
 */
export const SearchableList = <T extends unknown>({
  filterItems,
  allItems,
  renderItem,
  keyExtractor,
  searchDelayMs,
  headerComponent,
  testID,
  showsAllResultsOnEmptySearch = true,
}: SearchableListProps<T>) => {
  const {t} = useTranslation();
  const {theme} = useTheme();
  const styles = useStyles();

  const [searchTerm, setSearchTerm] = useState('');
  const [delayTimerId, setDelayTimerId] = useState<number>();
  const [filteredItems, setFilteredItems] = useState<T[]>([]);

  useMemo(() => {
    log.debug(`raw search term: '${searchTerm}'`);

    if (!searchTerm) {
      setDelayTimerId(undefined);
      setFilteredItems(showsAllResultsOnEmptySearch ? allItems : []);
      return;
    }

    const processedSearchTerm = searchTerm.trim().toLowerCase();

    log.debug('setting timer');
    setDelayTimerId(
      // @ts-ignore
      setTimeout(
        () => {
          setDelayTimerId(undefined);

          const filtered = filterItems(processedSearchTerm, allItems);
          log.debug(
            `Searching for '${processedSearchTerm}' returned ${filtered.length} results out of ${allItems.length}`,
          );

          setFilteredItems(filtered);
        },
        searchDelayMs !== undefined
          ? searchDelayMs
          : constants.explore.search.delayMs,
      ),
    );
  }, [
    allItems,
    filterItems,
    showsAllResultsOnEmptySearch,
    searchDelayMs,
    searchTerm,
  ]);

  return (
    <View style={styles.container} testID={testID}>
      <SearchBar
        placeholder={t('SEARCH_BAR_PLACEHOLDER')}
        numberOfLines={1}
        theme={theme}
        // This is a workaround for a bug where the characters already input are repeated on some Android keyboards
        // This disables predictive typing and autocorrect for Androids.
        //
        // See https://github.com/facebook/react-native/issues/30503
        keyboardType={
          Platform.OS === 'android' ? 'visible-password' : 'default'
        }
        round
        showLoading={delayTimerId !== undefined}
        onClear={() => {
          if (delayTimerId) {
            clearTimeout(delayTimerId);
          }
          setSearchTerm('');
        }}
        onChangeText={text => {
          if (delayTimerId) {
            clearTimeout(delayTimerId);
          }
          setSearchTerm(text);
        }}
        value={searchTerm}
      />
      <FlatList
        ListHeaderComponent={headerComponent}
        keyExtractor={keyExtractor}
        data={filteredItems}
        renderItem={({item}) => renderItem(item)}
      />
    </View>
  );
};

const useStyles = makeStyles(theme => ({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
}));
