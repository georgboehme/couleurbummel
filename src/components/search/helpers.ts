import log from '../../log';
import {CouleurbummelModel} from '../../model/model';
import {Address, Colours, Corporation} from '../../types/model';

export interface ColoursQuery {
  baseColourId: string;
  colourFamilyIds: string[];
}

const coloursMatch = (
  colours: Colours,
  colourIdsToFind: string[][],
  baseColourIdsToFind?: string[],
) => {
  // If we're searching for more or fewer colours than we have, it's not a match
  if (colourIdsToFind.length !== colours.colourIds.length) {
    return false;
  }

  // If the query contains a base colour, but the corporation doesn't have one or has the wrong one, it doesn't match.
  if (
    baseColourIdsToFind?.length &&
    (!colours.baseColourId ||
      baseColourIdsToFind.indexOf(colours.baseColourId) === -1) &&
    (!colours.percussionColourId ||
      baseColourIdsToFind.indexOf(colours.percussionColourId) === -1)
  ) {
    log.debug(
      'Base colour not matching',
      colours.baseColourId,
      baseColourIdsToFind,
    );
    return false;
  }

  let index = 0;
  let coloursFound = 0;

  // See if we can find the colours in the correct order.
  for (const i in colourIdsToFind) {
    const matchingColourIds = colourIdsToFind[i];

    if (
      // Blank colours are wildcards and always match
      !matchingColourIds.length ||
      matchingColourIds.indexOf(colours.colourIds[index]) !== -1
    ) {
      coloursFound++;
    }
    index++;
  }

  return coloursFound === colourIdsToFind.length;
};

export const searchColours = (
  query: ColoursQuery,
  data: [Address, Corporation][],
  model: CouleurbummelModel,
) => {
  log.debug('Serving query', query);

  const baseColourToSearch = model.colourFamily(
    query.baseColourId,
  )?.memberColourIds;

  const coloursToSearch = (query.colourFamilyIds || []).map(
    familyId => model.colourFamily(familyId)?.memberColourIds || [],
  );

  log.debug('Resolved colours', baseColourToSearch, coloursToSearch);

  return data.filter(([_address, corporation]) => {
    return (corporation.colours || []).some(cols =>
      coloursMatch(cols, coloursToSearch, baseColourToSearch),
    );
  });
};
