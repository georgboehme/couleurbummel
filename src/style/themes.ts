import {DarkTheme, DefaultTheme} from '@react-navigation/native';
import {Theme} from '@react-navigation/native/lib/typescript/src/types';
import {createTheme, darkColors, lightColors} from '@rneui/themed';
import {ColorValue, Platform} from 'react-native';

const ColourConstants = {
  primary: '#008040',
  primaryTransparent: 'rgba(0,128,64,0.75)',
  primaryDark: '#006633',
  accent: '#3CB371',
  accentTransparent: '#3CB37180',
  white: '#FFFFFF',
  whiteDarkModeText: '#E1E8EE',
  grey: '#EBEBEB',
  darkGrey: '#393E42FF',
  black: '#000000',
};

const sharedColours = {
  primary: ColourConstants.primary,
  primaryTransparent: ColourConstants.primaryTransparent,
  accent: ColourConstants.accent,
  accentTransparent: ColourConstants.accentTransparent,
  mapMarkerFill: ColourConstants.grey,
  mapMarkerCentreStroke: ColourConstants.black,
  iconOnPrimaryColourBackgroundColour: ColourConstants.whiteDarkModeText,
  fixed: {
    white: ColourConstants.white,
    lightGrey: ColourConstants.grey,
    darkGrey: ColourConstants.darkGrey,
    black: ColourConstants.black,
  },
};

const platformLightDefaults = Platform.select({
  default: lightColors.platform.android,
  ios: lightColors.platform.ios,
});

const platformDarkDefaults = Platform.select({
  default: darkColors.platform.android,
  ios: darkColors.platform.ios,
});

export const theme = createTheme({
  lightColors: {
    ...platformLightDefaults,
    ...sharedColours,
    text: ColourConstants.black,
  },
  darkColors: {
    ...platformDarkDefaults,
    ...sharedColours,
    text: ColourConstants.whiteDarkModeText,
  },
  components: {
    Icon: {type: 'material-community', color: ColourConstants.primary},
    ListItem: {
      containerStyle: {
        // Required to allow the selection highlight (defaults to theme background otherwise)
        backgroundColor: 'transparent',
      },
    },
    // Compiler warning fixed in https://github.com/react-native-elements/react-native-elements/pull/3691
    // @ts-ignore
    ListItemSwipeable: {
      containerStyle: {
        // Required to allow the selection highlight (defaults to theme background otherwise)
        backgroundColor: 'transparent',
      },
    },
    SearchBar: {
      containerStyle: {
        backgroundColor: 'transparent',
        borderTopColor: 'transparent',
        borderBottomColor: 'transparent',
      },
    },
  },
});

export const navigationTheme: {lightTheme: Theme; darkTheme: Theme} = {
  lightTheme: {
    dark: false,
    colors: {
      ...DefaultTheme.colors,
      primary: sharedColours.primary,
    },
  },
  darkTheme: {
    dark: true,
    colors: {
      ...DarkTheme.colors,
      primary: sharedColours.primary,
    },
  },
};

declare module '@rneui/themed' {
  export interface Colors {
    primaryTransparent: ColorValue;
    accent: ColorValue;
    accentTransparent: ColorValue;
    mapMarkerFill: ColorValue;
    mapMarkerCentreStroke: ColorValue;
    text: ColorValue;
    iconOnPrimaryColourBackgroundColour: ColorValue;
    fixed: {
      white: ColorValue;
      lightGrey: ColorValue;
      darkGrey: ColorValue;
      black: ColorValue;
    };
  }
}
