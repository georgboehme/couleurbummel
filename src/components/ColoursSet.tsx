import {makeStyles} from '@rneui/themed';
import React, {memo, useContext} from 'react';
import {ColorValue, View} from 'react-native';

import {Corporation} from '../types/model';
import Colours from './Colours';
import {ModelContext} from './contexts/ModelContext';

/**
 * A collection of one or more Colours views for a corporation
 */
export const ColoursSet = memo(({c}: {c: Corporation}) => {
  const styles = useStyles();
  const model = useContext(ModelContext);

  return (
    <View style={styles.coloursContainersContainer}>
      {Array.from(c?.colours || []).map((colourSet, idx) => {
        const cols =
          (colourSet.colourIds
            ?.map(col => model.colour(col))
            .filter(col => !!col) as ColorValue[]) || [];
        const percussion = model.colour(colourSet.percussionColourId);
        const baseColour = model.colour(colourSet.baseColourId);

        return (
          <View
            style={styles.coloursContainer}
            key={`list_colours_${c?.id}_${idx}`}>
            <Colours
              coloursKey={`colours_${c?.id}`}
              colours={cols}
              percussion={percussion}
              baseColour={baseColour}
            />
          </View>
        );
      })}
    </View>
  );
});

const useStyles = makeStyles(() => ({
  coloursContainersContainer: {
    width: '100%',
    marginVertical: 5,
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    maxHeight: 80,
    alignItems: 'center',
  },
  coloursContainer: {
    marginHorizontal: 5,
    aspectRatio: 3 / 2,
    maxWidth: 120,
    flexGrow: 1,
    flexShrink: 1,
  },
}));
