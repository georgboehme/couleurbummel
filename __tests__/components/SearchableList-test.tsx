import {ThemeProvider} from '@rneui/themed';
import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';
import {Text} from 'react-native';

import {SearchableList} from '../../src/components/SearchableList';
import {theme} from '../../src/style/themes';
import {waitForPromisesToResolveWithAct} from '../__helpers__/helpers';

describe('Country list', () => {
  const cityNames = ['Aachen', 'Berlin', 'München', 'Tübingen'];
  const testSearchableList = (
    <ThemeProvider theme={theme}>
      <SearchableList
        allItems={cityNames}
        filterItems={(searchTerm, items) =>
          items.filter(item => item.includes(searchTerm))
        }
        keyExtractor={item => item}
        renderItem={item => <Text>{item}</Text>}
      />
    </ThemeProvider>
  );

  it('renders correctly with no optional fields set', async () => {
    render(testSearchableList);

    await waitForPromisesToResolveWithAct().then(() =>
      expect(screen.toJSON()).toMatchSnapshot(),
    );
  });

  it('displays nothing if search term is empty', async () => {
    render(testSearchableList);

    await waitForPromisesToResolveWithAct().then(() =>
      expect(screen.toJSON()).toMatchSnapshot(),
    );
  });

  it('filters the list when searched', async () => {
    render(testSearchableList);

    const searchField = screen.getByPlaceholderText('SEARCH_BAR_PLACEHOLDER');
    fireEvent.changeText(searchField, 'chen');

    await waitForPromisesToResolveWithAct().then(() => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
