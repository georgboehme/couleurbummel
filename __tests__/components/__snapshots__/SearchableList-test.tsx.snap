// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country list displays nothing if search term is empty 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "transparent",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#dcdce1",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 15,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#86939e",
                        "fontSize": 18,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="SEARCH_BAR_PLACEHOLDER"
          placeholderTextColor="#86939e"
          platform="default"
          style={
            {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    data={
      [
        "Aachen",
        "Berlin",
        "München",
        "Tübingen",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Aachen
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Berlin
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          München
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Tübingen
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Country list filters the list when searched 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "transparent",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#dcdce1",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 15,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#86939e",
                        "fontSize": 18,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="SEARCH_BAR_PLACEHOLDER"
          placeholderTextColor="#86939e"
          platform="default"
          style={
            {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
          value="chen"
        />
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          >
            <ActivityIndicator
              style={
                {
                  "marginRight": 5,
                }
              }
            />
            <View
              style={
                {
                  "overflow": "hidden",
                }
              }
              testID="RNE__ICON__CONTAINER"
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="RNE__ICON__CONTAINER_ACTION"
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                  testID="RNE__ICON"
                >
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        {
                          "backgroundColor": "transparent",
                        },
                        {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                    testID="RNE__ICON__Component"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    data={
      [
        "Aachen",
        "Berlin",
        "München",
        "Tübingen",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Aachen
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Berlin
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          München
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Tübingen
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Country list renders correctly with no optional fields set 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "transparent",
        "borderBottomColor": "transparent",
        "borderBottomWidth": 1,
        "borderTopColor": "transparent",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#dcdce1",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 15,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#86939e",
                        "fontSize": 18,
                      },
                      {
                        "backgroundColor": "transparent",
                      },
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="SEARCH_BAR_PLACEHOLDER"
          placeholderTextColor="#86939e"
          platform="default"
          style={
            {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    data={
      [
        "Aachen",
        "Berlin",
        "München",
        "Tübingen",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Aachen
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Berlin
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          München
        </Text>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <Text>
          Tübingen
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
